File: app/src/main/java/com/example/locationmocker/domain/track/NaturalRunCursor.kt
package com.example.locationmocker.domain.track

import com.example.locationmocker.domain.model.PlaybackMode
import com.example.locationmocker.domain.model.RoutePoint
import com.example.locationmocker.domain.model.RouteSample
import com.example.locationmocker.domain.route.RouteMath
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.sin
import kotlin.random.Random

class NaturalRunCursor(
    route: List<RoutePoint>,
    private val mode: PlaybackMode,
    private val baseSpeedKmh: Float = 8.5f,
    private val updateIntervalMs: Long = 1_000L,
    private val random: Random = Random(System.currentTimeMillis()),
) {
    private val points = closeRoute(route)
    private val cumulativeMeters = buildCumulativeMeters(points)
    private val totalMeters = cumulativeMeters.lastOrNull() ?: 0.0
    private var traveledMeters = 0.0
    private var direction = 1
    private var completed = false
    private var lastLap = -1
    private var lapSpeedOffsetKmh = 0.0
    private var lapLaneBiasMeters = 0.0
    private var lapDriftAmplitudeMeters = 1.5
    private var lapPhase = 0.0
    @Volatile
    private var currentBaseSpeedKmh = baseSpeedKmh.coerceIn(6f, 12f)
    private var smoothedSpeedMetersPerSecond = (baseSpeedKmh / 3.6).toDouble()

    fun updateBaseSpeedKmh(speedKmh: Float) {
        currentBaseSpeedKmh = speedKmh.coerceIn(6f, 12f)
    }

    fun next(): RouteSample? {
        if (points.size < 2 || totalMeters <= 0.0 || completed) return null

        val lap = floor(traveledMeters / totalMeters).toInt()
        if (lap != lastLap) {
            lastLap = lap
            lapSpeedOffsetKmh = random.nextDouble(-0.45, 0.45)
            lapLaneBiasMeters = random.nextDouble(-0.9, 0.9)
            lapDriftAmplitudeMeters = random.nextDouble(1.0, 2.4)
            lapPhase = random.nextDouble(0.0, 2.0 * PI)
        }

        val currentDistance = normalizeDistance(traveledMeters)
        val progress = currentDistance / totalMeters
        val basePoint = interpolateAt(currentDistance)
        val nextDistance = normalizeDistance(currentDistance + direction * 2.0)
        val nextBasePoint = interpolateAt(nextDistance)
        val driftedPoint = applySmoothDrift(basePoint, nextBasePoint, progress)
        val lookAheadDistance = normalizeDistance(nextDistance + direction * 2.0)
        val nextProgress = lookAheadDistance / totalMeters
        val nextDriftedPoint = applySmoothDrift(nextBasePoint, interpolateAt(lookAheadDistance), nextProgress)
        val speed = smoothedSpeedMetersPerSecond(progress)
        val bearing = RouteMath.bearingDegrees(driftedPoint, nextDriftedPoint)

        advance(speed)
        return RouteSample(driftedPoint, speed, bearing)
    }

    private fun advance(speedMetersPerSecond: Float) {
        val stepMeters = speedMetersPerSecond * (updateIntervalMs / 1_000f)
        when (mode) {
            PlaybackMode.Once -> {
                traveledMeters += stepMeters
                if (traveledMeters >= totalMeters) completed = true
            }

            PlaybackMode.Loop -> {
                traveledMeters += stepMeters
            }

            PlaybackMode.PingPong -> {
                traveledMeters += stepMeters * direction
                when {
                    traveledMeters >= totalMeters -> {
                        traveledMeters = totalMeters
                        direction = -1
                    }

                    traveledMeters <= 0.0 -> {
                        traveledMeters = 0.0
                        direction = 1
                    }
                }
            }
        }
    }

    private fun smoothedSpeedMetersPerSecond(progress: Double): Float {
        val wave = sin(progress * 2.0 * PI * 3.0 + lapPhase) * 0.35 +
            sin(progress * 2.0 * PI * 7.0 + lapPhase * 0.7) * 0.15
        val speedKmh = (currentBaseSpeedKmh + lapSpeedOffsetKmh + wave).coerceIn(6.0, 12.0)
        val target = speedKmh / 3.6
        val alpha = (updateIntervalMs / 1_500.0).coerceIn(0.08, 0.28)
        smoothedSpeedMetersPerSecond += (target - smoothedSpeedMetersPerSecond) * alpha
        return smoothedSpeedMetersPerSecond.toFloat()
    }

    private fun applySmoothDrift(point: RoutePoint, nextPoint: RoutePoint, progress: Double): RoutePoint {
        val projection = TrackRoutePlanner.LocalProjection(point)
        val next = projection.project(nextPoint)
        val length = kotlin.math.sqrt(next.x * next.x + next.y * next.y).takeIf { it > 0.01 } ?: 1.0
        val normalX = -next.y / length
        val normalY = next.x / length
        val drift = (
            lapLaneBiasMeters +
                lapDriftAmplitudeMeters * sin(progress * 2.0 * PI * 2.0 + lapPhase) +
                0.45 * sin(progress * 2.0 * PI * 7.0 + lapPhase * 0.6)
            ).coerceIn(-3.0, 3.0)
        return projection.unproject(
            TrackRoutePlanner.ProjectedPoint(
                x = normalX * drift * direction,
                y = normalY * drift * direction,
            ),
        )
    }

    private fun interpolateAt(distanceMeters: Double): RoutePoint {
        val distance = distanceMeters.coerceIn(0.0, totalMeters)
        val segmentIndex = cumulativeMeters
            .indexOfLast { it <= distance }
            .coerceIn(0, cumulativeMeters.lastIndex - 1)
        val startDistance = cumulativeMeters[segmentIndex]
        val endDistance = cumulativeMeters[segmentIndex + 1]
        val fraction = if (endDistance == startDistance) {
            0.0
        } else {
            (distance - startDistance) / (endDistance - startDistance)
        }
        return RouteMath.interpolate(points[segmentIndex], points[segmentIndex + 1], fraction)
    }

    private fun normalizeDistance(distanceMeters: Double): Double {
        if (mode == PlaybackMode.PingPong) return distanceMeters.coerceIn(0.0, totalMeters)
        val remainder = distanceMeters % totalMeters
        return if (remainder < 0.0) remainder + totalMeters else remainder
    }

    private fun closeRoute(route: List<RoutePoint>): List<RoutePoint> {
        if (route.size < 2) return route
        return if (RouteMath.distanceMeters(route.first(), route.last()) <= 0.5) {
            route
        } else {
            route + route.first()
        }
    }

    private fun buildCumulativeMeters(points: List<RoutePoint>): List<Double> {
        if (points.isEmpty()) return emptyList()
        val result = mutableListOf(0.0)
        points.zipWithNext().forEach { (a, b) ->
            result += result.last() + RouteMath.distanceMeters(a, b)
        }
        return result
    }
}

File: LocationMocker/LocationMocker/Domain/Track/NaturalRunCursor.swift
import Foundation

final class NaturalRunCursor {
    private let points: [RoutePoint]
    private let mode: PlaybackMode
    private let updateIntervalMs: Int
    private let cumulativeMeters: [Double]
    private let totalMeters: Double

    private var traveledMeters = 0.0
    private var direction = 1
    private var completed = false
    private var lastLap = -1
    private var lapSpeedOffsetKmh = 0.0
    private var lapLaneBiasMeters = 0.0
    private var lapDriftAmplitudeMeters = 1.5
    private var lapPhase = 0.0
    private var currentBaseSpeedKmh: Double
    private var smoothedSpeedMps = 0.0

    init(route: [RoutePoint], mode: PlaybackMode, baseSpeedKmh: Float = 8.5, updateIntervalMs: Int = 1000) {
        self.points = Self.closeRoute(route)
        self.mode = mode
        self.currentBaseSpeedKmh = Double(min(max(baseSpeedKmh, 6), 12))
        self.updateIntervalMs = updateIntervalMs

        var cum = [0.0]
        for i in 0..<(self.points.count - 1) {
            cum.append(cum.last! + RouteMath.distanceMeters(self.points[i], self.points[i + 1]))
        }
        self.cumulativeMeters = cum
        self.totalMeters = cum.last ?? 0
        self.smoothedSpeedMps = currentBaseSpeedKmh / 3.6
    }

    func updateBaseSpeedKmh(_ speedKmh: Float) {
        currentBaseSpeedKmh = Double(min(max(speedKmh, 6), 12))
    }

    func next() -> RouteSample? {
        guard points.count >= 2, totalMeters > 0, !completed else { return nil }

        let lap = Int(floor(traveledMeters / totalMeters))
        if lap != lastLap {
            lastLap = lap
            lapSpeedOffsetKmh = Double.random(in: -0.45...0.45)
            lapLaneBiasMeters = Double.random(in: -0.9...0.9)
            lapDriftAmplitudeMeters = Double.random(in: 1.0...2.4)
            lapPhase = Double.random(in: 0...(2 * .pi))
        }

        let currentDist = normalizeDistance(traveledMeters)
        let progress = currentDist / totalMeters
        let basePoint = interpolateAt(currentDist)
        let nextDist = normalizeDistance(currentDist + Double(direction) * 2)
        let nextBasePoint = interpolateAt(nextDist)
        let drifted = applySmoothDrift(basePoint, next: nextBasePoint, progress: progress)

        let lookAhead = normalizeDistance(nextDist + Double(direction) * 2)
        let lookProgress = lookAhead / totalMeters
        let lookDrifted = applySmoothDrift(nextBasePoint, next: interpolateAt(lookAhead), progress: lookProgress)

        let speed = smoothedSpeed(progress)
        let bearing = RouteMath.bearingDegrees(drifted, lookDrifted)

        advance(speed)
        return RouteSample(point: drifted, speedMetersPerSecond: speed, bearingDegrees: bearing)
    }

    func reset() {
        traveledMeters = 0
        direction = 1
        completed = false
        lastLap = -1
        smoothedSpeedMps = currentBaseSpeedKmh / 3.6
    }

    private func advance(_ speedMps: Float) {
        let step = Double(speedMps) * (Double(updateIntervalMs) / 1000)
        switch mode {
        case .once:
            traveledMeters += step
            if traveledMeters >= totalMeters { completed = true }
        case .loop:
            traveledMeters += step
        case .pingPong:
            traveledMeters += step * Double(direction)
            if traveledMeters >= totalMeters { traveledMeters = totalMeters; direction = -1 }
            else if traveledMeters <= 0 { traveledMeters = 0; direction = 1 }
        }
    }

    private func smoothedSpeed(_ progress: Double) -> Float {
        let wave = sin(progress * 2 * .pi * 3 + lapPhase) * 0.35 +
            sin(progress * 2 * .pi * 7 + lapPhase * 0.7) * 0.15
        let speedKmh = min(max(currentBaseSpeedKmh + lapSpeedOffsetKmh + wave, 6), 12)
        let target = speedKmh / 3.6
        let alpha = min(max(Double(updateIntervalMs) / 1500.0, 0.08), 0.28)
        smoothedSpeedMps += (target - smoothedSpeedMps) * alpha
        return Float(smoothedSpeedMps)
    }

    private func applySmoothDrift(_ point: RoutePoint, next: RoutePoint, progress: Double) -> RoutePoint {
        let proj = TrackRoutePlanner.LocalProjection(origin: point)
        let np = proj.project(next)
        let length = sqrt(np.x * np.x + np.y * np.y)
        let len = length > 0.01 ? length : 1.0
        let nx = -np.y / len
        let ny = np.x / len

        let drift = min(max(
            lapLaneBiasMeters +
            lapDriftAmplitudeMeters * sin(progress * 2 * .pi * 2 + lapPhase) +
            0.45 * sin(progress * 2 * .pi * 7 + lapPhase * 0.6),
            -3), 3)
        return proj.unproject(TrackRoutePlanner.ProjectedPoint(
            x: nx * drift * Double(direction),
            y: ny * drift * Double(direction)
        ))
    }

    private func interpolateAt(_ distanceMeters: Double) -> RoutePoint {
        let d = min(max(distanceMeters, 0), totalMeters)
        let idx = cumulativeMeters.lastIndex(where: { $0 <= d }) ?? 0
        let segIdx = min(idx, cumulativeMeters.count - 2)
        let start = cumulativeMeters[segIdx]
        let end = cumulativeMeters[segIdx + 1]
        let frac = end == start ? 0 : (d - start) / (end - start)
        return RouteMath.interpolate(points[segIdx], points[segIdx + 1], fraction: frac)
    }

    private func normalizeDistance(_ d: Double) -> Double {
        if mode == .pingPong { return min(max(d, 0), totalMeters) }
        let rem = d.truncatingRemainder(dividingBy: totalMeters)
        return rem < 0 ? rem + totalMeters : rem
    }

    private static func closeRoute(_ route: [RoutePoint]) -> [RoutePoint] {
        guard route.count >= 2 else { return route }
        if RouteMath.distanceMeters(route.first!, route.last!) <= 0.5 { return route }
        return route + [route.first!]
    }
}
